runtime/hsa-runtime/core/runtime/amd_aie_aql_queue.cpp

```diff
@@ -41,28 +41,56 @@
 ////////////////////////////////////////////////////////////////////////////////
 
 #include "core/inc/amd_aie_aql_queue.h"
+#include "core/inc/amd_xdna_driver.h"
 
 #ifdef __linux__
 #include <fcntl.h>
 #include <sys/mman.h>
 #include <sys/stat.h>
-#include <sys/syscall.h>
-#include <unistd.h>
+#include <sys/ioctl.h>
 #endif
 
 #ifdef _WIN32
 #include <Windows.h>
 #endif
 
-#include <stdio.h>
-#include <string.h>
-#include <thread>
+#include <cstring>
 
 #include "core/inc/queue.h"
 #include "core/inc/runtime.h"
 #include "core/inc/signal.h"
 #include "core/util/utils.h"
 
+// The number of arguments in the packet payload before we start passing operands
+constexpr uint32_t NON_OPERAND_COUNT = 6;
+
+// Used to transform an address into a device address
+constexpr uint32_t DEV_ADDR_BASE = 0x04000000;
+constexpr uint32_t DEV_ADDR_OFFSET_MASK = 0x02FFFFFF;
+
+// The driver places a structure before each command in a command chain.
+// Need to increase the size of the command by the size of this structure.
+// In the following xdna driver source can see where this is implemented:
+// Commit hash: eddd92c0f61592c576a500f16efa24eb23667c23
+// https://github.com/amd/xdna-driver/blob/main/src/driver/amdxdna/aie2_msg_priv.h#L387-L391
+// https://github.com/amd/xdna-driver/blob/main/src/driver/amdxdna/aie2_message.c#L637
+constexpr uint32_t CMD_COUNT_SIZE_INCREASE = 3;
+
+// The size of an instruction in bytes
+constexpr uint32_t INSTR_SIZE_BYTES = 4;
+
+// Index of command payload where the instruction sequence
+// address is located
+constexpr uint32_t CMD_PKT_PAYLOAD_INSTRUCTION_SEQUENCE_IDX = 2;
+constexpr uint32_t CMD_PKT_PAYLOAD_INSTRUCTION_SEQUENCE_SIZE_IDX = 4;
+
+// Macro to concatenate two 32-bits addresses into a 64-bit address
+#define CONCAT_ADDR(HIGH, LOW) ((static_cast<uint64_t>(HIGH) << 32) | static_cast<uint64_t>(LOW))
```

Comment:
Better to avoid macros and use a function. Also if something like this does not already exist in utils.h, please add it there in as a general function.

Reply:
And make it constexpr ;)

Reply:
In c393105f70cdd24f4c3e81f10fc5111c28c197df I moved this into `util.h` doesn't seem like the other helper functions are constexpr so tried to be consistent with those. Let me know if you think that should change.

Reply:
`constexpr` is more of a property of the function so not all of them may be. A good rule of thump is that if a function can be evaluated at compile time and is inline, it should be constexpr.